### PR TITLE
Cuda11.4

### PR DESCRIPTION
**Pull request by oliviajain**
**Description**: Update GPU Pipelines and Dockerfiles to CUDA 11.4

**Motivation and Context**
Align with CUDA 11.4

### CODE REVIEW

File: tools/ci_build/github/azure-pipelines/orttraining-py-packaging-pipeline-torch181-cuda111.yml [hunk 1]
```diff
@@ -5,7 +5,7 @@ stages:
   parameters:
     build_py_parameters: --enable_training --update --build
     torch_version: '1.8.1'
-    cuda_version: '11.1'
+    cuda_version: '11.4'
```
Comment (liqunfu) [inline tools/ci_build/github/azure-pipelines/orttraining-py-packaging-pipeline-torch181-cuda111.yml:8]: for training we need to publish wheels for different Cuda versions. As its name suggested, this pipeline is for Cuda 11.1. So please keep this file unchanged.

File: tools/ci_build/github/azure-pipelines/orttraining-py-packaging-pipeline-torch190-cuda111.yml [hunk 2]
```diff
@@ -5,7 +5,7 @@ stages:
   parameters:
     build_py_parameters: --enable_training --update --build
     torch_version: '1.9.0'
-    cuda_version: '11.1'
+    cuda_version: '11.4'
```
Comment (liqunfu) [inline tools/ci_build/github/azure-pipelines/orttraining-py-packaging-pipeline-torch190-cuda111.yml:8]: Please undo this change. It is needed for Cuda 11.1 support.

**Review by liqunfu — CHANGES_REQUESTED** [review]

File: tools/ci_build/github/azure-pipelines/templates/orttraining-linux-gpu-ortmodule-test-ci-pipeline.yml [hunk 3]
```diff
@@ -33,7 +33,7 @@ steps:
...
+        bash -c "python3 -m pip uninstall -y -r /onnxruntime_src/tools/ci_build/github/linux/docker/scripts/training/requirements.txt && python3 -m pip install -r /onnxruntime_src/tools/ci_build/github/linux/docker/scripts/training/ortmodule/stage1/requirements_torch1.9.0_cu11.4.txt && python3 -m pip install -r /onnxruntime_src/tools/ci_build/github/linux/docker/scripts/training/ortmodule/stage2/requirements.txt && rm -rf /build/onnxruntime/ && python3 -m pip install /build/dist/onnxruntime*.whl && python3 -m onnxruntime.training.ortmodule.torch_cpp_extensions.install && /build/launch_test.py --cmd_line_with_args 'python orttraining_ortmodule_tests.py --mnist /mnist --bert_data /bert_data/hf_data/glue_data/CoLA/original/raw --transformers_cache /hf_models_cache/huggingface/transformers' --cwd /build" \
```
Comment (jywu-msft) [inline tools/ci_build/github/azure-pipelines/templates/orttraining-linux-gpu-ortmodule-test-ci-pipeline.yml:36]: probably don't touch this file either.  avoid touching training related files.

File: tools/ci_build/github/azure-pipelines/linux-gpu-ci-pipeline.yml [hunk 4]
```diff
@@ -13,7 +13,7 @@ jobs:
...
       Dockerfile: tools/ci_build/github/linux/docker/Dockerfile.manylinux2014_cuda11
       Context: tools/ci_build/github/linux/docker
-      DockerBuildArgs: "--network=host --build-arg POLICY=manylinux2014 --build-arg PLATFORM=x86_64 --build-arg BASEIMAGE=nvcr.io/nvidia/cuda:11.1.1-cudnn8-devel-centos7 --build-arg DEVTOOLSET_ROOTPATH=/opt/rh/devtoolset-9/root --build-arg PREPEND_PATH=/opt/rh/devtoolset-9/root/usr/bin: --build-arg LD_LIBRARY_PATH_ARG=/opt/rh/devtoolset-9/root/usr/lib64:/opt/rh/devtoolset-9/root/usr/lib:/opt/rh/devtoolset-9/root/usr/lib64/dyninst:/opt/rh/devtoolset-9/root/usr/lib/dyninst:/usr/local/lib64 --build-arg BUILD_UID=$( id -u )"
+      DockerBuildArgs: "--network=host --build-arg POLICY=manylinux2014 --build-arg PLATFORM=x86_64 --build-arg BASEIMAGE=nvcr.io/nvidia/cuda:11.4.1-cudnn8-devel-centos8 --build-arg DEVTOOLSET_ROOTPATH=/opt/rh/devtoolset-9/root --build-arg PREPEND_PATH=/opt/rh/devtoolset-9/root/usr/bin: --build-arg LD_LIBRARY_PATH_ARG=/opt/rh/devtoolset-9/root/usr/lib64:/opt/rh/devtoolset-9/root/usr/lib:/opt/rh/devtoolset-9/root/usr/lib64/dyninst:/opt/rh/devtoolset-9/root/usr/lib/dyninst:/usr/local/lib64 --build-arg BUILD_UID=$( id -u )"
```
[the remark this answers is not in the document]
Comment (jywu-msft) [reply tools/ci_build/github/azure-pipelines/linux-gpu-ci-pipeline.yml:16]: might be a typo. should be centos7

File: tools/ci_build/github/azure-pipelines/c-api-noopenmp-packaging-pipelines.yml [hunk 5]
```diff
@@ -40,14 +40,14 @@ jobs:
...
         Dockerfile: tools/ci_build/github/linux/docker/Dockerfile.manylinux2014_cuda11
         Context: tools/ci_build/github/linux/docker
-        DockerBuildArgs: "--network=host --build-arg POLICY=manylinux2014 --build-arg PLATFORM=x86_64 --build-arg BASEIMAGE=nvcr.io/nvidia/cuda:11.1-cudnn8-devel-centos7 --build-arg DEVTOOLSET_ROOTPATH=/opt/rh/devtoolset-9/root --build-arg PREPEND_PATH=/opt/rh/devtoolset-9/root/usr/bin: --build-arg LD_LIBRARY_PATH_ARG=/opt/rh/devtoolset-9/root/usr/lib64:/opt/rh/devtoolset-9/root/usr/lib:/opt/rh/devtoolset-9/root/usr/lib64/dyninst:/opt/rh/devtoolset-9/root/usr/lib/dyninst:/usr/local/lib64 --build-arg BUILD_UID=$( id -u )"
+        DockerBuildArgs: "--network=host --build-arg POLICY=manylinux2014 --build-arg PLATFORM=x86_64 --build-arg BASEIMAGE=nvcr.io/nvidia/cuda:11.4.1-cudnn8-devel-centos8 --build-arg DEVTOOLSET_ROOTPATH=/opt/rh/devtoolset-9/root --build-arg PREPEND_PATH=/opt/rh/devtoolset-9/root/usr/bin: --build-arg LD_LIBRARY_PATH_ARG=/opt/rh/devtoolset-9/root/usr/lib64:/opt/rh/devtoolset-9/root/usr/lib:/opt/rh/devtoolset-9/root/usr/lib64/dyninst:/opt/rh/devtoolset-9/root/usr/lib/dyninst:/usr/local/lib64 --build-arg BUILD_UID=$( id -u )"
```
Comment (jywu-msft) [inline tools/ci_build/github/azure-pipelines/c-api-noopenmp-packaging-pipelines.yml:50]: we need to stay on centos7 as @snnn mentioned

File: tools/ci_build/github/azure-pipelines/linux-gpu-ci-pipeline.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (oliviajain) [reply tools/ci_build/github/azure-pipelines/linux-gpu-ci-pipeline.yml:16]: I was looking at [these options](https://ngc.nvidia.com/catalog/containers/nvidia:cuda/tags)

11.4.1-runtime-centos7 (11.4.1/centos7/runtime/Dockerfile)
11.4.1-devel-centos7 (11.4.1/centos7/devel/Dockerfile)
11.4.1-base-centos7 (11.4.1/centos7/base/Dockerfile)

11.4.1-cudnn8-runtime-centos8 (11.4.1/centos8/runtime/cudnn8/Dockerfile)
11.4.1-runtime-centos8 (11.4.1/centos8/runtime/Dockerfile)
11.4.1-cudnn8-devel-centos8 (11.4.1/centos8/devel/cudnn8/Dockerfile)
11.4.1-devel-centos8 (11.4.1/centos8/devel/Dockerfile)
11.4.1-base-centos8 (11.4.1/centos8/base/Dockerfile)

[the remark this answers is not in the document]
Comment (snnn) [reply tools/ci_build/github/azure-pipelines/linux-gpu-ci-pipeline.yml:16]: Then they stopped supporting it.  I can add centos 8 support to our manylinux scripts though.

[the remark this answers is not in the document]
Comment (jywu-msft) [reply tools/ci_build/github/azure-pipelines/linux-gpu-ci-pipeline.yml:16]: no let's not switch yet.
we found something similar when updating the TensorRT manylinux dockerfile. even though there is no cudnn8 in the name, the build succeeds https://github.com/microsoft/onnxruntime/blob/master/tools/ci_build/github/linux/docker/Dockerfile.manylinux2014_cuda11_4_tensorrt8_0#L1
I think cudnn is in there.
actually, why don't we use the same one referenced in the trt8 manylinux dockerfile?

or update everything to 11.4.1-devel-centos7 as another option

File: tools/ci_build/github/azure-pipelines/nuget/templates/gpu.yml [hunk 6]
```diff
@@ -8,7 +8,7 @@ jobs:
     AgentPool : 'onnxruntime-gpu-winbuild'
     ArtifactName: 'drop-nuget'
     JobName: 'Windows_CI_GPU_CUDA_Dev'
-    BuildCommand: --build_dir $(Build.BinariesDirectory) --skip_submodule_sync --build_shared_lib --enable_onnx_tests --use_telemetry --cmake_generator "Visual Studio 16 2019" --use_cuda --cuda_version=11.4 --cuda_home="C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.4" --cudnn_home="C:\local\cudnn-11.4-windows-x64-v8.0.5.39\cuda" --cmake_extra_defines "CMAKE_CUDA_ARCHITECTURES=37;50;52;60;61;70;75;80"
+    BuildCommand: --build_dir $(Build.BinariesDirectory) --skip_submodule_sync --build_shared_lib --enable_onnx_tests --use_telemetry --cmake_generator "Visual Studio 16 2019" --use_cuda --cuda_version=11.4 --cuda_home="C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.4" --cudnn_home="C:\local\cudnn-11.4-windows-x64-v8.2.2.26\cuda " --cmake_extra_defines "CMAKE_CUDA_ARCHITECTURES=37;50;52;60;61;70;75;80"
```
Comment (snnn) [inline tools/ci_build/github/azure-pipelines/nuget/templates/gpu.yml:11]: extra space in cudnn_home

File: tools/ci_build/github/azure-pipelines/orttraining-py-packaging-pipeline-torch190-cuda114.yml [hunk 7]
```diff
@@ -5,7 +5,7 @@ stages:
   parameters:
     build_py_parameters: --enable_training --update --build
     torch_version: '1.9.0'
-    cuda_version: '11.4'
+    cuda_version: '11.1'
```
Comment (liqunfu) [inline tools/ci_build/github/azure-pipelines/orttraining-py-packaging-pipeline-torch190-cuda114.yml:8]: this one shall not be changed. it is meant to be 11.4

File: tools/ci_build/github/azure-pipelines/linux-gpu-ci-pipeline.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (oliviajain) [reply tools/ci_build/github/azure-pipelines/linux-gpu-ci-pipeline.yml]: nvcr.io/nvidia/cuda:11.4.0-devel-centos7 doesn't have cudnn but 
nvidia/cuda:11.4.0-cudnn8-devel-centos7 does.

File: tools/ci_build/github/azure-pipelines/c-api-noopenmp-packaging-pipelines.yml [hunk 8]
```diff
@@ -40,14 +40,14 @@ jobs:
...
         Dockerfile: tools/ci_build/github/linux/docker/Dockerfile.manylinux2014_cuda11
         Context: tools/ci_build/github/linux/docker
-        DockerBuildArgs: "--network=host --build-arg POLICY=manylinux2014 --build-arg PLATFORM=x86_64 --build-arg BASEIMAGE=nvcr.io/nvidia/cuda:11.1-cudnn8-devel-centos7 --build-arg DEVTOOLSET_ROOTPATH=/opt/rh/devtoolset-9/root --build-arg PREPEND_PATH=/opt/rh/devtoolset-9/root/usr/bin: --build-arg LD_LIBRARY_PATH_ARG=/opt/rh/devtoolset-9/root/usr/lib64:/opt/rh/devtoolset-9/root/usr/lib:/opt/rh/devtoolset-9/root/usr/lib64/dyninst:/opt/rh/devtoolset-9/root/usr/lib/dyninst:/usr/local/lib64 --build-arg BUILD_UID=$( id -u )"
+        DockerBuildArgs: "--network=host --build-arg POLICY=manylinux2014 --build-arg PLATFORM=x86_64 --build-arg BASEIMAGE=nvcr.io/nvidia/cuda:11.4.0-devel-centos7 --build-arg DEVTOOLSET_ROOTPATH=/opt/rh/devtoolset-9/root --build-arg PREPEND_PATH=/opt/rh/devtoolset-9/root/usr/bin: --build-arg LD_LIBRARY_PATH_ARG=/opt/rh/devtoolset-9/root/usr/lib64:/opt/rh/devtoolset-9/root/usr/lib:/opt/rh/devtoolset-9/root/usr/lib64/dyninst:/opt/rh/devtoolset-9/root/usr/lib/dyninst:/usr/local/lib64 --build-arg BUILD_UID=$( id -u )"
```
Comment (oliviajain) [inline tools/ci_build/github/azure-pipelines/c-api-noopenmp-packaging-pipelines.yml]: check pipeline - cudnn not found?

File: tools/ci_build/github/azure-pipelines/linux-gpu-ci-pipeline.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (jywu-msft) [reply tools/ci_build/github/azure-pipelines/linux-gpu-ci-pipeline.yml]: can you use nvidia/cuda:11.4.0-devel-centos7   ?  
use the exact same as in https://github.com/microsoft/onnxruntime/blob/master/tools/ci_build/github/linux/docker/Dockerfile.manylinux2014_cuda11_4_tensorrt8_0#L1 I mentioned above.
That is driving the Linux-TensorRT-CI pipeline, which seems to be working fine.

[the remark this answers is not in the document]
Comment (jywu-msft) [reply tools/ci_build/github/azure-pipelines/linux-gpu-ci-pipeline.yml]: can you compare that dockerfile and the one you are trying to update? I'm confused why there would be a failure. 
they should be almost the same (other than installing TensorRT dependencies)

[the remark this answers is not in the document]
Comment (oliviajain) [reply tools/ci_build/github/azure-pipelines/linux-gpu-ci-pipeline.yml]: Yes, 
- pulling nvcr.io/nvidia/cuda:11.4.0-devel-centos7 --> cudnn.h not found
- nvidia/cuda:11.4.0-devel-centos7 --> cudnn.h not found
-  nvidia/cuda:11.4.0-cudnn8-devel-centos7 --> ./usr/include/cudnn.h

[the remark this answers is not in the document]
Comment (snnn) [reply tools/ci_build/github/azure-pipelines/linux-gpu-ci-pipeline.yml]: They shouldn't have this difference. Did I miss something?

File: tools/ci_build/github/azure-pipelines/c-api-noopenmp-packaging-pipelines.yml [hunk 9]
```diff
@@ -40,14 +40,14 @@ jobs:
...
         Dockerfile: tools/ci_build/github/linux/docker/Dockerfile.manylinux2014_cuda11
         Context: tools/ci_build/github/linux/docker
-        DockerBuildArgs: "--network=host --build-arg POLICY=manylinux2014 --build-arg PLATFORM=x86_64 --build-arg BASEIMAGE=nvcr.io/nvidia/cuda:11.1-cudnn8-devel-centos7 --build-arg DEVTOOLSET_ROOTPATH=/opt/rh/devtoolset-9/root --build-arg PREPEND_PATH=/opt/rh/devtoolset-9/root/usr/bin: --build-arg LD_LIBRARY_PATH_ARG=/opt/rh/devtoolset-9/root/usr/lib64:/opt/rh/devtoolset-9/root/usr/lib:/opt/rh/devtoolset-9/root/usr/lib64/dyninst:/opt/rh/devtoolset-9/root/usr/lib/dyninst:/usr/local/lib64 --build-arg BUILD_UID=$( id -u )"
+        DockerBuildArgs: "--network=host --build-arg POLICY=manylinux2014 --build-arg PLATFORM=x86_64 --build-arg BASEIMAGE=nvidia/cuda:11.4.0-cudnn8-devel-centos7 --build-arg DEVTOOLSET_ROOTPATH=/opt/rh/devtoolset-9/root --build-arg PREPEND_PATH=/opt/rh/devtoolset-9/root/usr/bin: --build-arg LD_LIBRARY_PATH_ARG=/opt/rh/devtoolset-9/root/usr/lib64:/opt/rh/devtoolset-9/root/usr/lib:/opt/rh/devtoolset-9/root/usr/lib64/dyninst:/opt/rh/devtoolset-9/root/usr/lib/dyninst:/usr/local/lib64 --build-arg BUILD_UID=$( id -u )"
```
Comment (snnn) [inline tools/ci_build/github/azure-pipelines/c-api-noopenmp-packaging-pipelines.yml:50]: Please help me change "devtoolset-9" to "devtoolset-10" also.  Then they will be consistent with the CPU images. 
Otherwise your build could fail because recently I just updated the manylinux build scripts.

File: tools/ci_build/github/azure-pipelines/linux-gpu-ci-pipeline.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (snnn) [reply tools/ci_build/github/azure-pipelines/linux-gpu-ci-pipeline.yml]: So the idea is to get base image from nvidia/cuda:11.4.0-devel-centos7  then install cudnn by following the instructions at https://docs.nvidia.com/deeplearning/cudnn/install-guide/index.html#package-manager-rhel-install ?

[the remark this answers is not in the document]
Comment (jywu-msft) [reply tools/ci_build/github/azure-pipelines/linux-gpu-ci-pipeline.yml]: yes, looks like we'll have to install cudnn ourselves.
Dockerfile.manylinux2014_cuda11_4_tensorrt8_0  works because when it installs tensorrt packages, it will install the cudnn8 dependency.


Installing:
 libnvinfer-devel            x86_64     8.0.1-1.cuda11.3         cuda     157 M
 libnvinfer-plugin-devel     x86_64     8.0.1-1.cuda11.3         cuda      12 M
 libnvinfer-plugin8          x86_64     8.0.1-1.cuda11.3         cuda      12 M
 libnvinfer8                 x86_64     8.0.1-1.cuda11.3         cuda     139 M
 libnvonnxparsers-devel      x86_64     8.0.1-1.cuda11.3         cuda     415 k
 libnvonnxparsers8           x86_64     8.0.1-1.cuda11.3         cuda     713 k
 libnvparsers-devel          x86_64     8.0.1-1.cuda11.3         cuda     617 k
 libnvparsers8               x86_64     8.0.1-1.cuda11.3         cuda     850 k
Installing for dependencies:
 cuda-nvrtc-11-3             x86_64     11.3.109-1               cuda      34 M
 cuda-nvrtc-devel-11-3       x86_64     11.3.109-1               cuda      28 k
 libcudnn8                   x86_64     8.2.2.26-1.cuda11.4      cuda     769 M
 libcudnn8-devel             x86_64     8.2.2.26-1.cuda11.4      cuda     583 M

**Comment by oliviajain** [comment]
testing linux gpu ci then will resolve conflicts and run all https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=468660&view=logs&j=7536d2cd-87d4-54fe-4891-bfbbf2741d83&t=275bff32-624c-5245-5ea6-1890c77a0b18

**Comment by oliviajain** [comment]
> [testing linux gpu ci ](https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=468660&view=logs&j=7536d2cd-87d4-54fe-4891-bfbbf2741d83&t=275bff32-624c-5245-5ea6-1890c77a0b18)then will resolve conflicts and run all

https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=468660&view=logs&j=f2f63060-d9d6-52d0-adee-b97db5a9ab91&t=9885cde9-979c-51fc-3d1c-c8ad8dc15817&l=221. Build succeeded, but test is failing due to not finding cuda11.4

File: tools/ci_build/github/windows/setup_env_trt.bat [hunk 10]
```diff
@@ -1,2 +1,2 @@
-set PATH=C:\azcopy;C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.1\bin;C:\local\cudnn-11.4-windows-x64-v8.2.2.26\cuda\bin;%PATH%
+set PATH=C:\azcopy;C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.4\bin;C:\local\cudnn-11.4-windows-x64-v8.2.2.26\cuda\bin;%PATH%
```
Comment (jywu-msft) [inline tools/ci_build/github/windows/setup_env_trt.bat:1]: can you revert this setup_env_trt.bat change for now?
this will allow the unit tests to pass.

Comment (snnn) [reply tools/ci_build/github/windows/setup_env_trt.bat:1]: Ah! I think I got it. This TRT bat is for TRT pipelines. And the windows TRT build machine may not have the latest driver. (I haven't checked). Then it would be ok if you use 11.1 on it. But once you change it to 11.4, it could cause problems.

Comment (snnn) [reply tools/ci_build/github/windows/setup_env_trt.bat:1]: @oliviajain , if you have time, you can make a quick change and set the machine pool to 'Win-GPU-2019', and queue a build. While you are waiting the result, you can follow @jywu-msft 's advice doing the other part.

Comment (snnn) [reply tools/ci_build/github/windows/setup_env_trt.bat:1]: I apologize I didn't make it clear.

**Review by snnn — DISMISSED** [review]

File: tools/ci_build/github/azure-pipelines/templates/py-packaging-stage.yml [hunk 11]
```diff
@@ -21,7 +21,7 @@ parameters:
   default: false
 
 - name: enable_linux_gpu_training_cu111
-  displayName: 'Whether Linux GPU Cuda 11.1 package is built.'
+  displayName: 'Whether Linux GPU Cuda 11.4 package is built.'
```
Comment (snnn) [inline tools/ci_build/github/azure-pipelines/templates/py-packaging-stage.yml:24]: Revert this line of change.

**Comment by oliviajain** [comment]
All passes checked - kicked off packaging pipelines.

File: tools/ci_build/github/windows/setup_env_trt.bat (hunk 10 above)
Comment (jywu-msft) [reply tools/ci_build/github/windows/setup_env_trt.bat]: > Ah! I think I got it. This TRT bat is for TRT pipelines. And the windows TRT build machine may not have the latest driver. (I haven't checked). Then it would be ok if you use 11.1 on it. But once you change it to 11.4, it could cause problems.

FYI. it's not about the driver. there's an issue with using CUDA 11.4 on Windows which causes 2 unit tests to fail. we reported to Nvidia. They reproduced the issue and are looking into it.

**Comment by snnn** [comment]
Python packaging pipeline Windows GPU: cudnn_home='C:\local\cudnn-11.4-windows-x64-v8.0.5.39\cuda' valid=False

**Comment by oliviajain** [comment]
Linux GPU Python 38
\cuda_home='/usr/local/cuda-11.4' valid=False. cudnn_home='/usr/local/cuda-11.4' valid=False

**Comment by snnn** [comment]
Let me try it again.

**Comment by oliviajain** [comment]
I [validated ](https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=469418&view=logs&j=7536d2cd-87d4-54fe-4891-bfbbf2741d83&t=275bff32-624c-5245-5ea6-1890c77a0b18&l=22 )linux-gpu-ci-pipeline with **nvidia/cuda:11.4.0-cudnn8-devel-centos7 and therefore not installing cudnn in dockerfile**. Would we prefer this over nvcr.io/nvidia/cuda:11.4.0-devel-centos7 ?

**Comment by oliviajain** [comment]
> Let me try it again.

worked

**Comment by jywu-msft** [comment]
> I [validated ](https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=469418&view=logs&j=7536d2cd-87d4-54fe-4891-bfbbf2741d83&t=275bff32-624c-5245-5ea6-1890c77a0b18&l=22)linux-gpu-ci-pipeline with **nvidia/cuda:11.4.0-cudnn8-devel-centos7 and therefore not installing cudnn in dockerfile**. Would we prefer this over nvcr.io/nvidia/cuda:11.4.0-devel-centos7 + cudnn yum install?

yes, it's preferred we don't need to install anything ourselves.

**Comment by oliviajain** [comment]
> > I [validated ](https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=469418&view=logs&j=7536d2cd-87d4-54fe-4891-bfbbf2741d83&t=275bff32-624c-5245-5ea6-1890c77a0b18&l=22)linux-gpu-ci-pipeline with **nvidia/cuda:11.4.0-cudnn8-devel-centos7 and therefore not installing cudnn in dockerfile**. Would we prefer this over nvcr.io/nvidia/cuda:11.4.0-devel-centos7 + cudnn yum install?
> 
> yes, it's preferred we don't need to install anything ourselves.

Let me run the CIs with that image and if it passes, I'll kick off packaging pipelines then hopefully we can merge it in.

**Review by snnn — APPROVED** [review]

**Review by jywu-msft — APPROVED** [review]

**Review by liqunfu — APPROVED** [review]
